### PR TITLE
feat: add Python 3.12 support

### LINKED ISSUE #446: Support for Python 3.12
I do not expect any changes needed in `cabinetry` to support Python 3.12. Currently not all backends of `pyhf` support that version, so official support in `cabinetry` should follow `pyhf`: https://github.com/scikit-hep/pyhf/issues/2372.

### PR DESCRIPTION
**Pull request by matthewfeickert**
Resolves #446 

```
* docs: Add Python 3.12 PyPI trove classifier
* ci: Add Python 3.12 to CI
* ci: Add workflow dispatch trigger to CI
* ci: List installed Python packages
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
The error that is currently being seen from `iminuit` is due to https://github.com/scikit-hep/iminuit/pull/977#issuecomment-2136430074, but is getting caused by the pre-release of `numpy` being picked up from https://github.com/scikit-hep/hist/issues/570. So this is a compound issue that would be fixed by a new `hist` release, but also shows that a new `iminuit` release would be nice.

**Comment by matthewfeickert** [comment]
@alexander-held this should be ready for review now given that [`hist` `v2.7.3`](https://github.com/scikit-hep/hist/releases/tag/v2.7.3) is released now.

File: .github/workflows/test-package.yml [hunk 1]
```diff
@@ -19,10 +19,10 @@ jobs:
       with:
         fetch-depth: 0
 
-    - name: Set up Python 3.11
+    - name: Set up Python
       uses: actions/setup-python@v5
       with:
-        python-version: '3.11'
+        python-version: '3.x'
```
Comment (alexander-held) [inline .github/workflows/test-package.yml:25]: How does the "3.x" here work?

Comment (matthewfeickert) [reply .github/workflows/test-package.yml:25]: `3.x` resolves to the latest Python that `actions/setup-python` supports. For packaging you're not testing anything about `cabinetry` itself other than it builds, and that is very luckily not brittle to CPython releases in terms of breaking metadata changes.

c.f. https://github.com/actions/setup-python/blob/82c7e631bb3cdc910f68e0081d67478d79c6982d/docs/advanced-usage.md#specifying-a-python-version for the technical details.

Comment (alexander-held) [reply .github/workflows/test-package.yml:25]: Thanks, I learned something new!

**Review by alexander-held — APPROVED** [review]
Thanks, this all looks good to me!

related previous PR for 3.11 was #400

```
* add Python 3.12 PyPI trove classifier
* add Python 3.12 to CI
* add workflow dispatch trigger to CI
* list installed Python packages in CI
```

**Pull request merged**